tree/ntuple/v7/test/ntuple_processor_join.cxx

```diff
@@ -0,0 +1,279 @@
+         auto fldJ = model->MakeField<int>("j");
+         auto fldK = model->MakeField<int>("k");
+         auto fldA = model->MakeField<float>("a");
+         auto ntuple = RNTupleWriter::Recreate(std::move(model), fNTupleNames[3], fFileNames[3]);
+
+         for (const auto &i : {3, 14, 9, 8, 11, 10, 0, 13, 1, 5, 6, 12, 2, 4, 7}) {
+            *fldI = i;
+            *fldJ = *fldI / 2;
+            *fldK = *fldJ / 2;
+            if (*fldI % 2 == 1) {
+               *fldA = *fldI * 0.1f;
+               ntuple->Fill();
+            }
+         }
+      }
+   }
+
+   void TearDown() override
+   {
+      for (const auto &fileName : fFileNames) {
+         std::remove(fileName.c_str());
+      }
+   }
+};
+
+TEST_F(RNTupleJoinProcessorTest, Basic)
+{
+   std::vector<RNTupleOpenSpec> ntuples;
+   try {
+      auto proc = RNTupleProcessor::CreateJoin(ntuples, {});
+      FAIL() << "creating a processor without at least one RNTuple should throw";
+   } catch (const RException &err) {
+      EXPECT_THAT(err.what(), testing::HasSubstr("at least one RNTuple must be provided"));
+   }
+
+   ntuples = {{fNTupleNames[0], fFileNames[0]}};
+
+   int nEntries = 0;
+   auto proc = RNTupleProcessor::CreateJoin(ntuples, {});
+   for (const auto &entry : *proc) {
+      auto i = entry.GetPtr<int>("i");
+      EXPECT_EQ(proc->GetNEntriesProcessed(), *i);
+      ++nEntries;
+   }
+   EXPECT_EQ(nEntries, 10);
+   EXPECT_EQ(nEntries, proc->GetNEntriesProcessed());
+}
+
+TEST_F(RNTupleJoinProcessorTest, Aligned)
+{
+   try {
+      std::vector<RNTupleOpenSpec> ntuples = {{fNTupleNames[0], fFileNames[0]}, {fNTupleNames[0], fFileNames[0]}};
+      auto proc = RNTupleProcessor::CreateJoin(ntuples, {});
+      FAIL() << "ntuples with the same name cannot be joined horizontally";
+   } catch (const RException &err) {
+      EXPECT_THAT(err.what(), testing::HasSubstr("horizontal joining of RNTuples with the same name is not allowed"));
+   }
+
+   std::vector<RNTupleOpenSpec> ntuples = {{fNTupleNames[1], fFileNames[1]}, {fNTupleNames[2], fFileNames[2]}};
+
+   auto proc = RNTupleProcessor::CreateJoin(ntuples, {});
+
+   std::vector<float> yExpected;
+
+   int nEntries = 0;
+   for (auto &entry : *proc) {
+      auto i = entry.GetPtr<int>("i");
+
+      yExpected = {static_cast<float>(*i * 0.2), 3.14, static_cast<float>(*i * 1.3)};
+      EXPECT_EQ(yExpected, *entry.GetPtr<std::vector<float>>("y"));
+
+      EXPECT_FLOAT_EQ(*i * 2.f, *entry.GetPtr<float>("ntuple3.z"));
+
+      ++nEntries;
+   }
+   EXPECT_EQ(nEntries, 5);
+   EXPECT_EQ(nEntries, proc->GetNEntriesProcessed());
+}
+
+TEST_F(RNTupleJoinProcessorTest, IdenticalFieldNames)
+{
+   std::vector<RNTupleOpenSpec> ntuples = {{fNTupleNames[1], fFileNames[1]}, {fNTupleNames[2], fFileNames[2]}};
+
+   auto proc = RNTupleProcessor::CreateJoin(ntuples, {});
+
+   auto i = proc->GetEntry().GetPtr<int>("i");
+   for (auto &entry : *proc) {
+      EXPECT_EQ(*i, *entry.GetPtr<int>("ntuple3.i"));
+   }
```

Comment:
This test will succeed if the processor things they are no entry.  Should we add a check of the number of entries seen?